### PR TITLE
Resolve for typedefs/aliases and records in `typeArguments` & resolve duplicate typedefs

### PR DESCRIPTION
**Pull request by dickermoshe**
Closes: https://github.com/dart-lang/mockito/issues/775 
Closes: https://github.com/dart-lang/mockito/issues/777 
Closes: https://github.com/dart-lang/mockito/issues/778

The fix for these issues is closely related so I've done both of them in a single PR.

- Mockito wasn't checking for typedefs/aliases in `typeArguments`
- it also wasn't looking into records on type args. 
- Typedefs which had the same underlying type were ignored even if the typedef had a different name

These issues cause the dreaded `is missing from the asset URI mapping` error.

See the related issue for a repro of each of these. I've also added regression tests for all of these.

Looking forward to a swift review!

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by srawlins** [comment]
CC @mosuem

**Comment by dickermoshe** [comment]
It seems that the bulk of this code is rely on `InterfaceType` as the only interface object, while records are interfaces themselves.
It seems that the `addTypesFrom` function, which until now only took a `InterfaceType` will now have to take records too.

**Comment by dickermoshe** [comment]
Could this be cherrypicked into a 5.4.5 release?

**Comment by dickermoshe** [comment]
CI is broken on the entire repo as dart stable does not support `analyzer 6.9`. Tests pass locally on my machine.

**Comment by dickermoshe** [comment]
@mosuem 
I know people hate they get this, but.... Ping

**Comment by mosuem** [comment]
I don't mind at all, but I am also not a maintainer of this repo and maybe not the right person to ping ;) I just stumbled across the same CI error in https://github.com/dart-lang/test/pull/2293

**Comment by mosuem** [comment]
Maybe @srawlins wants to take a look (I hope he you don't mind the ping). I am fine refiling the PR to move to `dart-lang/test` and merging this first.

File: test/builder/auto_mocks_test.dart
Comment (srawlins) [inline test/builder/auto_mocks_test.dart:3681]: Hmm, not 'typedefs', I think you mean 'type arguments'?

File: test/builder/auto_mocks_test.dart
Comment (srawlins) [inline test/builder/auto_mocks_test.dart:3705]: "nested type arguments"

File: test/builder/auto_mocks_test.dart
Comment (srawlins) [inline test/builder/auto_mocks_test.dart:3557]: nit: "it's" instead of "its"

Comment (srawlins) [inline test/builder/auto_mocks_test.dart:3557]: I'm not sure what you mean here by "a type parameter of function which returns another type." In this test case, the typedef appears to only be used as a _type argument_ in a class definition `class Foo extends BaseFoo<CreateBar>`.

File: test/builder/auto_mocks_test.dart
Comment (srawlins) [inline test/builder/auto_mocks_test.dart:3593]: nit: space between `T,` and `P`, as that is how the formatter would format it.

File: test/builder/auto_mocks_test.dart
Comment (srawlins) [inline test/builder/auto_mocks_test.dart:3611]: I don't see any assertions here related to `BarDef` or `BarDef2`.

File: test/builder/auto_mocks_test.dart
Comment (srawlins) [inline test/builder/auto_mocks_test.dart:3586]: Instead of "duplicate" type parameter, do you just mean some type parameter after the first parameter? You could s say something like "any type parameter."

File: lib/src/builder.dart
Comment (srawlins) [inline lib/src/builder.dart:184]: I think this can read better if we hoist these properties out with `if case`, instead of `addAll` and `if`:

```dart
if (type.element?.library case var library?) {
  librariesWithTypes.add(library);
}
if (type.alias?element.library case var library?) {
  librariesWithTypes.add(library);
}
```

The `var library?` syntax is still kind of weird to me, but I'm trying to use it more. Basically "if that is non-null then extract it to a variable named `library`."

File: CHANGELOG.md
Comment (srawlins) [inline CHANGELOG.md:12]: I think instead of "typedef-aliases," we generally call them "type aliases."

**Review by srawlins — CHANGES_REQUESTED** [review]
Thanks so much! I left a few notes but generally the fix is gold 😁

**Comment by srawlins** [comment]
@mosuem I love a friendly ping 😁 I am fine doing whatever procedure works best for you, migrating to dart-lang/test. I think a git commit can be wired up that attributes the fix to @dickermoshe still...

For external PRs like this, I think I've been able to just hand-copy the PR, once landed, into google3, and then re-export. But there might be a good copybara mechanism for that, at this point...

**Comment by mosuem** [comment]
I'll have to check the copybara settings, but normally even g3 SoT repos should be able to accept Github PRs. We can probably copy the config from package:intl, then it should just work.

**Comment by mosuem** [comment]
But please go ahead with merging this, any infra work can wait, features are much nicer ;)

File: CHANGELOG.md
Comment (dickermoshe) [reply CHANGELOG.md:12]: Oh yeah, where do you think I got it from? 😉😜
https://github.com/dart-lang/mockito/blob/f72791d3dd5cb2acc4f5c7da065ffe70d2ad1e0b/CHANGELOG.md?plain=1#L51-L60

I'll fix it anyway 😂

File: lib/src/builder.dart
Comment (dickermoshe) [reply lib/src/builder.dart:184]: Agree. Big fan of the `case` in `if` checks too.
It's a shame it isn't a well known feature.

File: test/builder/auto_mocks_test.dart
Comment (dickermoshe) [reply test/builder/auto_mocks_test.dart:3557]: How about?
`typedef mocks are generated properly when it\'s a function which returns any type`

In this instance, `Bar` was being ignored becuase the return type of `CreateBar` want being resolved

File: test/builder/auto_mocks_test.dart
Comment (dickermoshe) [reply test/builder/auto_mocks_test.dart:3611]: before this pr, the above test would throw an exception that `BarDef2` is not found, so the above is sufficient.
I'll add a small test to if you want me too.
```dart
expect(mocksContent, contains('_i2.BarDef get t1'));
expect(mocksContent, contains('_i2.BarDef2 get t2'));
```

**Comment by dickermoshe** [comment]
Thanks for the timely response. This is the coolest library ever, thank so much!
Fixed all the nits, lmk if there is anything else

File: test/builder/auto_mocks_test.dart
Comment (srawlins) [reply test/builder/auto_mocks_test.dart]: That SG, thanks!

**Comment by srawlins** [comment]
OK we should be good to go. Ship it!

**Review by srawlins — APPROVED** [review]

**Pull request merged**
